walk/walk.go

```diff
@@ -366,28 +338,72 @@ func shouldFollow(wc *walkConfig, dir, rel string, ent fs.DirEntry) bool {
 		// A symlink, but not one we should follow.
 		return false
 	}
-	if _, err := os.Stat(path.Join(dir, ent.Name())); err != nil {
+	if _, err := os.Stat(path.Join(ctx.rootDir, rel, ent.Name())); err != nil {
 		// A symlink, but not one we could resolve.
 		return false
 	}
 	return true
 }
 
+// Information lasting the lifetime of the fs walk
+type buildTrieContext struct {
+	rootDir           string
+	readBuildFilesDir string
+
+	// An error group to handle error propagation
+	eg      *errgroup.Group
+	limitCh chan struct{}
+}
+
 type pathTrie struct {
-	entry    fs.DirEntry
-	files    []fs.DirEntry
+	rel string
+
+	files    []string
 	children []*pathTrie
+
+	walkConfig   *walkConfig
+	build        *rule.File
+	buildFileErr error
+
+	rw *sync.RWMutex
 }
 
 // Basic factory method to ensure the entry is properly copied
-func newTrie(entry fs.DirEntry) *pathTrie {
-	return &pathTrie{
-		entry: entry,
-	}
+func (trie *pathTrie) addChild(c *pathTrie) *pathTrie {
+	trie.rw.Lock()
+	defer trie.rw.Unlock()
+
+	trie.children = append(trie.children, c)
+	return c
+}
+
+func (trie *pathTrie) addFiles(p []string) {
```

Comment:
Now that the `updateOnly` logic is part of the concurrent fs-traversal the construction of `trie.{children,files}` may cross goroutines and needs these lock util methods.

Note that this one adding files accepts an array so we lock once per directory, not once per file within the directory.